### ISSUE TITLE
Float values have error in rounding

### ISSUE THREAD

**Issue by kzangeli** [telefonicaid/fiware-orion#1073]
A recent test showed us that a number-attribute with the value 3.3 is presented as 3.2999999998 in a query.  In mongo 3.3 is stored. The rounding error comes after - probably in the snprintf phase in render (the toJson() method).

We need to investigate and try to fix this.

**Effort: 2 man day**

**MORE INFO**:
The JSON beautifier (python -mtool.json) behaves differently in different machines.

This attribute:

    "attr": 3.300000"

Is 'beautified' to

    "attr": 3.3

in Jenkins.  In my machine (CentOS 6.5). I get:

    "attr": 3.2999999999999998

This makes two functests fail:
  creation_with_attribute_as_object.test
  simple_creation.out

We need to see whether we can send an option to the beautifier to change its behaviour or, implement a new beautifier, selectable via parameter to orionCurl.
Completely change the beautifier doesn't seem like an option - too many --REGEXPECT-- to modify ...

**Comment by fgalan** [comment]
Some important information in PR #1071 comment about this issue:

> I just tested and I see 3.3 in mongo. Not 2.2999...
>
> That is good, at least. I imagine the rounding error is in the last step (snprintf to a string in the render phase). We need to investigate and see if we can do anything about it.

**Comment by fgalan** [comment]
@jmcanterafonseca Is there any indication about float precission in JSON or JavaScript specifications?

**Comment by fgalan** [comment]
Surprisingly, jenkins seems to render in the right way (thus making the test fail):

```
Orion Functional Test Log File:
================================================================================
Orion Functional tests starting
Thu Jul 16 13:53:59 CEST 2015

-----  POST Entities V2 with attribute as object  -----
(0970_create_entity/creation_with_attribute_as_object.test) output not as expected
VALIDATION ERROR: input line:
       "attr03": 3.3,
does not match ref line:
       "attr03": 3.2999999999999998,



-----  POST Entities V2  -----
(0970_create_entity/simple_creation.test) output not as expected
VALIDATION ERROR: input line:
       "attr03": 3.3,
does not match ref line:
       "attr03": 3.2999999999999998,


================================================================================

----------- Failing tests ------------------
  o  0970_create_entity/creation_with_attribute_as_object.test
  o  0970_create_entity/simple_creation.test
```

It is not clear why jenkins behaves different compared with the other environment where we have tested.

**Comment by kzangeli** [comment]
Seen in https://docs.python.org/2/tutorial/floatingpoint.html (last phrase):

In versions prior to Python 2.7 and Python 3.1, Python rounded this value to 17 significant digits, giving ‘0.10000000000000001’. In current versions, Python displays a value based on the shortest decimal fraction that rounds correctly back to the true binary value, resulting simply in ‘0.1’.

This is exactly what happens in my machine (I get 16 decimals).
I checked my python version and I have 2.6.6 (prior to 2.7).

Jenkins seems to have a newer version of python ...

So, the solution may simply be to make sure we all have a python newer than { 2.7, 3.1 }

**Comment by kzangeli** [comment]
Problem solved by simply installing a newer version of **python**.
Noy my centos behaves the same way Jenkins does.

If anybody encounters this problem, the solution is simply to install a newer version of python.

In PR #1082 the two failing functests are changed to "3.3" instead of "3.299999999999998"

**Comment by fgalan** [comment]
Fixed in PR #1082
